### PR TITLE
Added di composition snippet

### PR DESCRIPTION
**Pull request by thekiba**
Нужен перевод

### CODE REVIEW

File: snippets/di-composition.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+@Directive({ selector: '[item]' })
+export class ItemDirective implements OnInit, OnDestroy {
+
+  @HostBinding('class.selected') isActivated: boolean = false;
+
+  constructor(@Host() private linker: ItemsLinker) {}
+
+  @HostListener('click') onClick() {
```
Comment (kirjs) [inline snippets/di-composition.md]: This is a very elaborate example, also hostbinding and hostlistener each deserve their own snippet.

**Review by kirjs — CHANGES_REQUESTED** [review]

**Comment by fetis** [comment]
this described in Angular doc https://angular.io/guide/hierarchical-dependency-injection#component-level-injectors
do we need to have this as a snippet?

**Comment by thekiba** [comment]
@fetis yep, because this is not a popular way to using angular di

File: snippets/di-composition.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+@Component({
+  selector: '[provide]',
+  providers: [ Service ]
+})
+export class ProvideDirective {}
+```
+
+# Links
+https://stackblitz.com/edit/angular-cdk-happy-animals
+
```
Comment (fetis) [inline snippets/di-composition.md]:
```suggestion
https://angular.io/guide/hierarchical-dependency-injection#component-level-injectors

```

**Review by fetis — CHANGES_REQUESTED** [review]

**Comment by fetis** [comment]
btw, "DI composition" says nothing for me. Maybe re-phrase it to more clear name?

**Comment by kirjs** [comment]
Can we have a shorter example please

File: snippets/component-level-providers.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+@Component({
+  selector: 'provide',
+  template: '<ng-content></ng-content>',
+  providers: [ Service ]
+})
+export class ProvideComponent {}
+```
+
+```typescript
+@Component({
```
Comment (fetis) [inline snippets/component-level-providers.md]:
```suggestion
@Directive({
```

File: snippets/component-level-providers.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import { Injectable, Component, OnInit, OnDestroy, Host } from '@angular/core';
+
+@Injectable()
+export class ItemsLinker implements OnDestroy {
+  links: Set<ItemComponent> = new Set();
+
+  link(item: ItemComponent) {
+    this.links.add(item);
+  }
+  unlink(item: ItemComponent) {
```
Comment (fetis) [inline snippets/component-level-providers.md:49]:
```suggestion

  unlink(item: ItemComponent) {
```

File: snippets/component-level-providers.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+https://stackblitz.com/edit/angular-cdk-happy-animals
+
+# ComponentCode
+```typescript
+import { Injectable, Component, OnInit, OnDestroy, Host } from '@angular/core';
+
+@Injectable()
+export class ItemsLinker implements OnDestroy {
+  links: Set<ItemComponent> = new Set();
```
Comment (fetis) [inline snippets/component-level-providers.md]: just a better isolation
```suggestion
  private links: Set<ItemComponent> = new Set();
```

**Review by fetis — CHANGES_REQUESTED** [review]

**Review by fetis — APPROVED** [review]

**Review by kirjs — APPROVED** [review]

**Pull request merged**